### PR TITLE
feat: Implement Webview navigation

### PR DESCRIPTION
**Pull request by acezard**
Flagship app navigation feature
Implement cozy-bar side of AA cozy-bar webview navigation

- [x] Add or upgrade cozy-intent
- [x] Add or upgrade cozy-device-helper
- [x] Upgrade Node CI and .nvmrc to 16 if not already done
- [x] Fix eventual issues caused by the aforementioned upgrades
- [x] Implement AA state in Redux
- [x] Implement View according to state
- [x] Implement side-effects according to user actions
- [x] Add tests where possible

### CODE REVIEW

**Comment by trollepierre** [comment]
I don't understand why we compare with a release branch... ?

**Comment by Crash--** [comment]
Because master (aka v8.X) is currently unusable in our apps. We decided to not fix the issue during this projet. 

You can have a look at the issue when reading the work to do on the bar (check our roadmap and linked ... links) ;)

**Comment by trollepierre** [comment]
so I don't understand why we do have 6 commits of 6 topics different inside this PR.

**Comment by acezard** [comment]
> so I don't understand why we do have 6 commits of 6 topics different inside this PR.

Well, all the commits are related to the same feature. It is customary to not upgrade libraries in the same commit of feature code

File: test/lib/api.spec.jsx [hunk 1]
```diff
@@ -128,4 +128,14 @@ describe('api spec', function() {
...
+
+  it('should set webviewContext', function() {
+    const { setWebviewContext } = api
+    // default webviewContext is `undefined`
+    expect(store.getState().context.webviewContext).toBe(undefined)
+    setWebviewContext('foo')
+    expect(store.getState().context.webviewContext).toBe('foo')
+    setWebviewContext('bar')
+    expect(store.getState().context.webviewContext).toBe('bar')
+  })
```
Comment (trollepierre) [inline test/lib/api.spec.jsx:140]: after this test, let's setWebviewContext(undefined) again

File: src/styles/bar.css [hunk 2]
```diff
@@ -37,6 +37,10 @@
   [role=banner] .coz-bar-container {
     padding: 0 1em 0 0;
   }
+
+  [role=banner] > .coz-is-flagship-app .coz-bar-container {
+    padding: 0 1em 0 1em;
```
Comment (trollepierre) [inline src/styles/bar.css:42]:
```suggestion
    padding: 0 1em;
```

File: src/components/Bar.jsx [hunk 3]
```diff
@@ -218,9 +220,13 @@ export class Bar extends Component {
       ? { '--cozBarThemePrimaryContrastTextColor': pctColor }
       : {}
     const themeStyle = { ...pStyle, ...pctStyle }
+    const flagshipClass = isFlagshipApp ? ' coz-is-flagship-app' : ''
 
     return (
-      <div className={`coz-bar-wrapper coz-theme-${theme}`} style={themeStyle}>
+      <div
+        className={`coz-bar-wrapper coz-theme-${theme}${flagshipClass}`}
```
Comment (trollepierre) [inline src/components/Bar.jsx:227]: why not using classNames ?

File: src/components/Bar.jsx [hunk 4]
```diff
@@ -218,9 +220,13 @@ export class Bar extends Component {
...
+        style={themeStyle}
+      >
```
Comment (trollepierre) [inline src/components/Bar.jsx:229]: I would add one test, when isFlagship true, then we add correct class

File: src/index.spec.jsx [hunk 5]
```diff
@@ -1,7 +1,13 @@
 import CozyClient from 'cozy-client'
+import jestFetchMock from 'jest-fetch-mock'
+
 import * as cozyBar from './index'
 
 describe('init', () => {
+  beforeAll(() => {
+    global.fetch = jestFetchMock
+  })
```
Comment (trollepierre) [inline src/index.spec.jsx:9]: I don't expect to add this in the test. Why?

nit: I was expecting a test to check setWebviewContext is provided as a proxiedApi.

File: src/lib/reducers/context.js [hunk 6]
```diff
@@ -19,6 +20,10 @@ export const shouldEnableClaudy = state => {
   return !!claudyActions && !!claudyActions.length
 }
 
+export const getWebviewContext = state => {
+  return state.webviewContext
```
Comment (trollepierre) [inline src/lib/reducers/context.js:24]: export const getWebviewContext = state => state.webviewContext;

or

export const getWebviewContext = { webviewContext } => webviewContext;

**Review by trollepierre — CHANGES_REQUESTED** [review]

File: src/components/Apps/ButtonCozyHome.jsx [hunk 7]
```diff
@@ -7,7 +7,7 @@ export const ButtonCozyHome = ({ webviewContext, homeHref }) => {
     return (
       <button
         onClick={() => {
-          webviewContext.call('openApp', 'home')
+          webviewContext.call('backToHome')
```
Comment (trollepierre) [inline src/components/Apps/ButtonCozyHome.jsx:10]: let's test the argument provided here 🙏

File: src/components/Bar.jsx (hunk 3 above)
Comment (acezard) [reply src/components/Bar.jsx:227]: Well it's not in the dependencies, I didn't want to add any runtime dependency here as I wanted minimal changes

File: src/components/Bar.jsx (hunk 4 above)
Comment (acezard) [reply src/components/Bar.jsx:229]: Alright

File: src/index.spec.jsx (hunk 5 above)
Comment (acezard) [reply src/index.spec.jsx:9]: Tests were broken without that mock. Not sure what config is the problem but it seemed good enough for now as I'm not updating everything

File: src/lib/reducers/context.js (hunk 6 above)
Comment (acezard) [reply src/lib/reducers/context.js:24]: I agree but here I just followed the already existing code style for clarity, even if it's verbose

File: src/components/Apps/ButtonCozyHome.jsx (hunk 7 above)
Comment (acezard) [reply src/components/Apps/ButtonCozyHome.jsx:10]: Alright

File: test/lib/api.spec.jsx (hunk 1 above)
Comment (acezard) [reply test/lib/api.spec.jsx:140]: Alright

**Review by trollepierre — APPROVED** [review]
thank you :-)

**Review by trollepierre — APPROVED** [review]

**Review by trollepierre — APPROVED** [review]

**Pull request merged**